api-docs: Clarify narrow parameter in `/get-messages` endpoint.

laurynmm opened the pull request:
Clarifies the narrow parameter description by adding information about what a user's message history includes, about bots not being automatically subscribed to the owner's streams, and about the `streams:public` narrow.

Adds some of that information to the main description for the endpoint as well, and changes the link in the main description to go to the parameter instead of to the 'construct-a-narrow' page, which is linked to in the parameter description.

Fixes #19477.

---

**Notes**:
- Do we also want to update the '[construct-a-narrow](https://zulip.com/api/construct-narrow)' documentation to include something about either the `streams:public` narrow specifically or the results being limited to a user's message history and what that includes?
- Do we want to update other instances of narrow parameters to have a similar description? These seem to be:
  - [check if messages match narrow](https://zulip.com/api/check-messages-match-narrow)
  - [register event queue](https://zulip.com/api/register-queue#parameter-narrow)
  - [real time events API](https://zulip.com/api/real-time-events#parameter-narrow)

---

**Screenshots and screen captures:**

<details>
<summary>Get messages main description</summary>

[Current documentation](https://zulip.com/api/get-messages)
![Screenshot from 2022-10-26 17-48-58](https://user-images.githubusercontent.com/63245456/198077293-ca59c02f-0478-4225-9c55-7f9e54b415ae.png)
</details>

<details>
<summary>Get messages narrow parameter description</summary>

[Current documentation](https://zulip.com/api/get-messages#parameter-narrow)
![Screenshot from 2022-10-26 17-48-47](https://user-images.githubusercontent.com/63245456/198077321-53e316a2-13c6-4fef-af76-0961808897c5.png)
</details>

---

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

timabbott commented:
> Do we want to update other instances of narrow parameters to have a similar description?

Yeah, I think so. For /events and /register, we definitely hear from people wondering "Why don't I get any messages when testing this", because they haven't subscribed their bot to any streams. 

We probably want a multi-faceted fix for this -- e.g. we should potentially make the UI for creating bots offer to let you add them to streams, so that this class of problem doesn't come up; perhaps you can start a thread in #integrations or #design to discuss how that should work.

But I think having an explicit mention in https://zulip.com/api/real-time-events and on the /register documentation that a user will only receive messages sent to streams to which it is subscribed would be a very helpful reminder in helping eliminate this not infrequent case of folks feeling like our bots system doesn't work because they didn't know they needed to subscribe the bot to streams.

timabbott commented:
Do we also want to update the '[construct-a-narrow](https://zulip.com/api/construct-narrow)' documentation to include something about either the streams:public narrow specifically or the results being limited to a user's message history and what that includes?

I think the "construct a narrow" page should probably explain the algorithm clearly. It's a bit subtle -- so I think a user-facing summary of the rules in the `ok_to_include_history` comments that explain it for Zulip developers would be a very helpful section to add to that page.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -5086,6 +5089,18 @@ paths:
+            Bots are not automatically subscribed to the same streams as the
+            bot's creator.
+
+            If you would like to fetch messages from public streams instead of
```
"from all public streams" might be best here.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -5006,12 +5006,15 @@ paths:
+        message history, which means that the results will be from streams
+        that the user is subscribed to and from private/group messages in
+        which the user is a participant. Note that bots are not automatically
+        subscribed to the same streams as the bot's creator.
```
Do you think something like "Note that a user's personal message history does not contain messages sent to streams before they subscribe, and newly created bot users usually are not be subscribed to any streams." would be clearer here? And then perhaps we can swap the order of the paragraphs so that the "By specifying a narrow filter" paragraph comes just after might feel more natural. 

(I think you can via the API create a bot with initial stream subscriptions, and as noted elsewhere, I'm thinking we should probably add UI for that too).

laurynmm commented:
**Updates**:
- Revised the `/get-messages` documentation update based on Tim's feedback above.
- Added updates to the shared `Narrow` parameter that's used in both the `/register-queue` and `/real-time-events` documentation.
- Updated the main description for the `/check-messages-match-narrow` endpoint.

---

**Pending**:
- I'm working on the `/construct-a-narrow` update, but thought I would post these changes for review now. Also that could be a follow-up commit to these changes.
- CZO conversation about updating the UI for creating bots to have an option/ability to subscribe to streams.

---

**Updated screenshots**:

<details>
<summary>Get messages - main description</summary>

[Current documentation](https://zulip.com/api/get-messages)
![Screenshot from 2022-10-28 15-52-59](https://user-images.githubusercontent.com/63245456/198696213-05a30071-00ac-4264-a5b2-2e9e1ed52d5a.png)
</details>

<details>
<summary>Get messages - narrow parameter</summary>

[Current documentation](https://zulip.com/api/get-messages#parameter-narrow)
![Screenshot from 2022-10-28 15-53-15](https://user-images.githubusercontent.com/63245456/198696232-ba35efaf-82bd-4b4e-9cd6-1eb22415ee31.png)
</details>

<details>
<summary>Check if matches narrow - main description</summary>

[Current documentation](https://zulip.com/api/check-messages-match-narrow)
![Screenshot from 2022-10-28 15-53-26](https://user-images.githubusercontent.com/63245456/198696234-23fb4eb3-7d26-48fa-8c18-dc3bb28ded20.png)
</details>

<details>
<summary>Real time events and register queue - narrow parameter </summary>

[Current documentation](https://zulip.com/api/real-time-events#parameter-narrow)
![Screenshot from 2022-10-28 15-53-43](https://user-images.githubusercontent.com/63245456/198696237-68c8371b-e81b-4159-bb0e-02a1726818b8.png)
</details>

timabbott commented:
I think the note about narrows being limited to a user's message history feels a bit confusing for the events parameter -- since we're talking about processing of new messages there anyway. 

I didn't have a clear idea of how to express that, so ended up fiddling with this and ending up with the following diff. Key changes include:
* Correcting a misleading statement about how shared history searches work; it's quite complicated (see the comments in `ok_to_include_history`) but I found a /help/ article that explains it well enough to link to. Definitely an idea we'll want to cover clearly on the "Construct a narrow" page, which probably needs a larger pass as you've mentioned, so we can leave that for a follow-up.
* Rewriting the real-time-events pages description of /narrow to more distinctly frame things as "this is updates for new mesages"; I realized splitting the private messages and streams examples would be helpful here.

``` diff
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index ce08c12d52..841915a9f4 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -5119,18 +5119,22 @@ paths:
             The narrow where you want to fetch the messages from. See how to
             [construct a narrow](/api/construct-narrow).
 
-            Note that, in general, narrows are limited to the user's message
-            history, which means that results will be from streams that the
-            user is subscribed to and from private/group messages in which
-            the user is a participant.
+            Note that many narrows, including all that lack a `stream`
+            or `streams` operator, search the user's personal message
+            history. See [searching shared
+            history](/help/search-for-messages#searching-shared-history)
+            for details.
 
-            Newly created bot users are not usually subscribed to any streams.
-
-            If you would like to fetch messages from all public streams instead
+            For example, if you would like to fetch messages from all public streams instead
             of only the user's message history, then a specific narrow for
             messages sent to all public streams can be used:
             `{"operator": "streams", "operand": "public"}`.
 
+            Newly created bot users are not usually subscribed to any
+            streams, so bots using this API should either be
+            subscribed to appropriate streams or use a shared history
+            search narrow with this endpoint.
+
             **Changes**: In Zulip 2.1.0, added support for using user/stream IDs
             when constructing narrows for a message's sender, its stream and/or
             its recipient(s).
@@ -17046,18 +17050,28 @@ components:
         [narrow filter(s)](/api/construct-narrow) for which you'd
         like to receive events for.
 
-        For instance, to receive events for the stream
-        `Denmark`, you would specify `narrow=[["stream", "Denmark"]]`.
-        Another example is `narrow=[["is", "private"]]` for private
-        messages.
+        For example, to receive events for private messages (including
+        group private messages) received by the user, one can use
+        `narrow=[["is", "private"]]`.
+
+        Unlike the API for [fetching messages](/api/get-messages),
+        this narrow parameter is simply a filter on messages that the
+        user receives through their stream subscriptions (or because
+        they are a recipient of a private message).
+
+        This means that a client that requests a `narrow` filter of
+        `[["stream", "Denmark"]]` will receive events for new messages
+        sent to that stream while the user is subscribed to that
+        stream. The client will not receive any message events at all
+        if the user is not subscribed to `"Denmark"`.
 
-        Note that narrows are limited to the user's message history,
-        which does not contain messages sent to streams before they
-        [subscribe](/api/subscribe), and that newly created bot users
-        are not usually subscribed to any streams.
+        Newly created bot users are not usually subscribed to any
+        streams, so bots using this API need to be
+        [subscribed](/api/subscribe) to any streams whose messages
+        you'd like them to process using this endpoint.
 
-        There is a specific narrow for messages sent to all public
-        streams, which is `narrow=[["streams", "public"]]`.
+        See the `all_public_streams` parameter for how to process all
+        public stream messages in an organization.
 
         Defaults to `[]`.
       content:
```

timabbott commented:
Merged, with those changes; thanks so much for making this happen @laurynmm! This is a super tangly area, so I really appreciate your helping with this.
